### ISSUE TITLE
Unexpected token: expected `;`, found `#`.

### ISSUE THREAD

**Issue by Jerrody** [rust-lang/rust#106020]
```rust
#[cfg(feature = "validation")]
[debug_utils_ext, (*surface_extensions).to_owned()].concat()

#[cfg(not(feature = "validation"))]
surface_extensions.to_owned()
```

Error code:
```rust
error: expected `;`, found `#`
  --> src\engine\renderer\context\instance.rs:44:73
   |
44 |             [debug_utils_ext, (*surface_extensions).to_owned()].concat()
   |                                                                         ^ help: add `;` here
45 |
46 |             #[cfg(not(feature = "validation"))]
   |             - unexpected token
   ```
   
Metadata:
```
rustc 1.67.0-nightly (e1d819583 2022-12-05)
binary: rustc
commit-hash: e1d819583f0bf13b016b119c1c2c43e6d3979450
commit-date: 2022-12-05
host: x86_64-pc-windows-msvc
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by compiler-errors** [comment]
What is the issue you are reporting here @Jerrody ?

**Comment by Jerrody** [comment]
> What is the issue you are reporting here @Jerrody ?

Well, I thought that compiler can understand that `#[cfg()]` is related to the block of code, but it fails.

**Comment by Nilstrieb** [comment]
Can you post a minimal reproducible example that demonstrates your issue? Or at least mention exactly what you'd expect to happen and why.

**Comment by Jerrody** [comment]
> Can you post a minimal reproducible example that demonstrates your issue? Or at least mention exactly what you'd expect to happen and why.
> 
> Even code that is cfged out is still required to parse successfully.

Let's show you what exactly problem I have.
Full code:
```rust
        let instance_extensions = {
            let surface_extensions =
                ash_window::enumerate_required_extensions(window.raw_display_handle()).track()?;

            #[cfg(feature = "validation")]
            let debug_utils_ext = vec![ash::extensions::ext::DebugUtils::name().as_ptr()];

            #[cfg(feature = "validation")]
            [debug_utils_ext, (*surface_extensions).to_owned()].concat()

            #[cfg(not(feature = "validation"))]
            surface_extensions.to_owned()
        };

**Comment by SkiFire13** [comment]
You seem to already have posted this [in this discussion in the internals forum](https://internals.rust-lang.org/t/make-possible-to-use-procedural-macros-with-scopes/17778) a month ago. What you're asking seems a rather big change to me, I'm not that familiar with how the compiler parser works but I believe it currently requires your program syntax to be correct before `#[cfg(...)]` macros are expanded, which  is not true for your example, so some fundamental changes would be needed.

**Comment by Jerrody** [comment]
> You seem to already have posted this [in this discussion in the internals forum](https://internals.rust-lang.org/t/make-possible-to-use-procedural-macros-with-scopes/17778) a month ago. What you're asking seems a rather big change to me, I'm not that familiar with how the compiler parser works but I believe it currently requires your program syntax to be correct before `#[cfg(...)]` macros are expanded, which  is not true for your example, so some fundamental changes would be needed.

Oh, damn, sorry. I totally forgot that I made a post.
Well, if parser works like this it is bad, anyway, exists alternative way to write like this.

**Comment by lukas-code** [comment]
The workaround for this is to wrap the expressions in a block:

```rust
let instance_extensions = {
    let surface_extensions =
        ash_window::enumerate_required_extensions(window.raw_display_handle()).track()?;

    #[cfg(feature = "validation")]
    let debug_utils_ext = vec![ash::extensions::ext::DebugUtils::name().as_ptr()];

    #[cfg(feature = "validation")]
    { [debug_utils_ext, (*surface_extensions).to_owned()].concat() }

    #[cfg(not(feature = "validation"))]
    { surface_extensions.to_owned() }
};
```

**Comment by Jerrody** [comment]
> The workaround for this is to wrap the expressions in a block:
> 
> ```rust
> let instance_extensions = {
>     let surface_extensions =
>         ash_window::enumerate_required_extensions(window.raw_display_handle()).track()?;
> 
>     #[cfg(feature = "validation")]
>     let debug_utils_ext = vec![ash::extensions::ext::DebugUtils::name().as_ptr()];
> 
>     #[cfg(feature = "validation")]
>     { [debug_utils_ext, (*surface_extensions).to_owned()].concat() }
> 
>     #[cfg(not(feature = "validation"))]
>     { surface_extensions.to_owned() }
> };
> ```

I solved this issue exactly how you did.

**Comment by albertlarsan68** [comment]
Another way to do this is using the `cfg!` macro and `if ... else`.
I personally thinks it is more clear, and should optimize the same.

```rust
let instance_extensions = {
    let surface_extensions =
        ash_window::enumerate_required_extensions(window.raw_display_handle()).track()?;

    #[cfg(feature = "validation")]
    let debug_utils_ext = vec![ash::extensions::ext::DebugUtils::name().as_ptr()];

    if cfg!(feature = "validation") {
        [debug_utils_ext, (*surface_extensions).to_owned()].concat()
    } else {
        surface_extensions.to_owned()
    }
};
```

**Comment by Jerrody** [comment]
> Another way to do this is using the `cfg!` macro and `if ... else`. I personally thinks it is more clear, and should optimize the same.
> 
> ```rust
> let instance_extensions = {
>     let surface_extensions =
>         ash_window::enumerate_required_extensions(window.raw_display_handle()).track()?;
> 
>     #[cfg(feature = "validation")]
>     let debug_utils_ext = vec![ash::extensions::ext::DebugUtils::name().as_ptr()];
> 
>     if cfg!(feature = "validation") {
>         [debug_utils_ext, (*surface_extensions).to_owned()].concat()
>     } else {
>         surface_extensions.to_owned()
>     }
> };
> ```

Work is not the same, due to `cfg!` compile-time evaluating an expression that we pass in it and evaluates only it, but at the same time everything compiles what is inside of `if else`.
